### ISSUE TITLE
CosmosDiagnostics regressions:  ITrace currently lose all information on exceptions

### ISSUE THREAD

**Issue by ealsur** [Azure/azure-cosmos-dotnet-v3#2261]
On the latest master state, 408 exceptions coming from TaskCanceledException, have no traces/diagnotics.

## With nuget 3.16.0

Stack trace:
```
Microsoft.Azure.Cosmos.CosmosException : Response status code does not indicate success: RequestTimeout (408); Substatus: 0; ActivityId: ; Reason: (GatewayStoreClient Request Timeout. Start Time UTC:02/24/2021 00:31:03; Total Duration:25117.7439 Ms; Request Timeout 10000 Ms; Http Client Timeout:65000 Ms; Activity id: 8d65fd0d-b0a5-42bf-a365-58a57b6d472c;);
 ---> System.Threading.Tasks.TaskCanceledException: The operation was canceled.
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean allowHttp2, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Microsoft.Azure.Cosmos.CosmosHttpClientCore.ExecuteHttpHelperAsync(HttpRequestMessage requestMessage, ResourceType resourceType, CancellationToken cancellationToken)
   at Microsoft.Azure.Cosmos.CosmosHttpClientCore.SendHttpHelperAsync(Func`1 createRequestMessageAsync, ResourceType resourceType, CosmosDiagnosticsContext diagnosticsContext, HttpTimeoutPolicy timeoutPolicy, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Cosmos.ResponseMessage.EnsureSuccessStatusCode()
   at Microsoft.Azure.Cosmos.CosmosResponseFactoryCore.ProcessMessage[T](ResponseMessage responseMessage, Func`2 createResponse)
   at Microsoft.Azure.Cosmos.CosmosResponseFactoryCore.CreateItemResponse[T](ResponseMessage responseMessage)
   at Microsoft.Azure.Cosmos.ContainerCore.CreateItemAsync[T](CosmosDiagnosticsContext diagnosticsContext, T item, ITrace trace, Nullable`1 partitionKey, ItemRequestOptions requestOptions, CancellationToken cancellationToken)
   at Microsoft.Azure.Cosmos.ClientContextCore.RunWithDiagnosticsHelperAsync[TResult](CosmosDiagnosticsContext diagnosticsContext, ITrace trace, Func`3 task)
```

Diagnostics:

```
{
    "DiagnosticVersion": "2",
    "Summary": {
        "StartUtc": "2021-02-24T00:31:03.1949892Z",
        "RunningElapsedTimeInMs": 26038.6541,
        "UserAgent": "cosmos-netstandard-sdk/3.16.0|3.15.4|03|X64|Linux 4.14.35-linuxkit 1 SMP |.NET Core 3.1.12|",
        "TotalRequestCount": 3,
        "FailedRequestCount": 3,
        "Operation": "CosmosDiagnosticsContextCore"
    },
    "Context": [
        {
            "Id": "SendHttpHelperAsync:HttpTimeoutPolicyControlPlaneRetriableHotPath",
            "ElapsedTimeInMs": 5068.9725
        },
        {
            "Id": "PointOperationStatistics",
            "ActivityId": "8d65fd0d-b0a5-42bf-a365-58a57b6d472c",
            "ResponseTimeUtc": "2021-02-24T00:31:08.2644021Z",
            "StatusCode": 503,
            "SubStatusCode": 0,
            "RequestCharge": 0.0,
            "RequestUri": "https://sqlmergev3crud2102232205test1-eastus.documents-test.windows-int.net//addresses/?$resolveFor=dbs%2f8mWcBw%3d%3d%2fcolls%2f8mWcB9TARSg%3d%2fdocs&$filter=protocol eq rntbd&$partitionKeyRangeIds=52",
            "ErrorMessage": "System.Threading.Tasks.TaskCanceledException: The operation was canceled.\n   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean allowHttp2, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)\n   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)\n   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)\n   at Microsoft.Azure.Cosmos.CosmosHttpClientCore.ExecuteHttpHelperAsync(HttpRequestMessage requestMessage, ResourceType resourceType, CancellationToken cancellationToken)\n   at Microsoft.Azure.Cosmos.CosmosHttpClientCore.SendHttpHelperAsync(Func`1 createRequestMessageAsync, ResourceType resourceType, CosmosDiagnosticsContext diagnosticsContext, HttpTimeoutPolicy timeoutPolicy, CancellationToken cancellationToken)",
            "RequestSessionToken": null,
            "ResponseSessionToken": null
        },
        {
            "Id": "SendHttpHelperAsync:HttpTimeoutPolicyControlPlaneRetriableHotPath",
            "ElapsedTimeInMs": 5019.2667
        },
        {
            "Id": "PointOperationStatistics",
            "ActivityId": "8d65fd0d-b0a5-42bf-a365-58a57b6d472c",
            "ResponseTimeUtc": "2021-02-24T00:31:13.2844153Z",
            "StatusCode": 503,
            "SubStatusCode": 0,
            "RequestCharge": 0.0,
            "RequestUri": "https://sqlmergev3crud2102232205test1-eastus.documents-test.windows-int.net//addresses/?$resolveFor=dbs%2f8mWcBw%3d%3d%2fcolls%2f8mWcB9TARSg%3d%2fdocs&$filter=protocol eq rntbd&$partitionKeyRangeIds=52",
            "ErrorMessage": "System.Threading.Tasks.TaskCanceledException: The operation was canceled.\n   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean allowHttp2, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)\n   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)\n   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)\n   at Microsoft.Azure.Cosmos.CosmosHttpClientCore.ExecuteHttpHelperAsync(HttpRequestMessage requestMessage, ResourceType resourceType, CancellationToken cancellationToken)\n   at Microsoft.Azure.Cosmos.CosmosHttpClientCore.SendHttpHelperAsync(Func`1 createRequestMessageAsync, ResourceType resourceType, CosmosDiagnosticsContext diagnosticsContext, HttpTimeoutPolicy timeoutPolicy, CancellationToken cancellationToken)",
            "RequestSessionToken": null,
            "ResponseSessionToken": null
        },
        {
            "Id": "HttpRetryDelay; Delay:00:00:01 seconds; Current request timeout 00:00:05; TimeoutPolicy: HttpTimeoutPolicyControlPlaneRetriableHotPath",
            "ElapsedTimeInMs": 5010.3614
        },
        {
            "Id": "SendHttpHelperAsync:HttpTimeoutPolicyControlPlaneRetriableHotPath",
            "ElapsedTimeInMs": 10015.7708
        },
        {
            "Id": "PointOperationStatistics",
            "ActivityId": "8d65fd0d-b0a5-42bf-a365-58a57b6d472c",
            "ResponseTimeUtc": "2021-02-24T00:31:28.3117112Z",
            "StatusCode": 503,
            "SubStatusCode": 0,
            "RequestCharge": 0.0,
            "RequestUri": "https://sqlmergev3crud2102232205test1-eastus.documents-test.windows-int.net//addresses/?$resolveFor=dbs%2f8mWcBw%3d%3d%2fcolls%2f8mWcB9TARSg%3d%2fdocs&$filter=protocol eq rntbd&$partitionKeyRangeIds=52",
            "ErrorMessage": "System.Threading.Tasks.TaskCanceledException: The operation was canceled.\n   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean allowHttp2, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)\n   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)\n   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)\n   at Microsoft.Azure.Cosmos.CosmosHttpClientCore.ExecuteHttpHelperAsync(HttpRequestMessage requestMessage, ResourceType resourceType, CancellationToken cancellationToken)\n   at Microsoft.Azure.Cosmos.CosmosHttpClientCore.SendHttpHelperAsync(Func`1 createRequestMessageAsync, ResourceType resourceType, CosmosDiagnosticsContext diagnosticsContext, HttpTimeoutPolicy timeoutPolicy, CancellationToken cancellationToken)",
            "RequestSessionToken": null,
            "ResponseSessionToken": null
        }
    ]
}
```

## With current master

Stacktrace:

```
Microsoft.Azure.Cosmos.CosmosException : Response status code does not indicate success: RequestTimeout (408); Substatus: 0; ActivityId: ; Reason: (GatewayStoreClient Request Timeout. Start Time UTC:02/25/2021 05:53:18; Total Duration:18753.9153 Ms; Request Timeout 10000 Ms; Http Client Timeout:65000 Ms; Activity id: ad399267-09f2-463e-9929-ff61022fedbf;);
 ---> System.Threading.Tasks.TaskCanceledException: The operation was canceled.
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean allowHttp2, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Microsoft.Azure.Cosmos.CosmosHttpClientCore.ExecuteHttpHelperAsync(HttpRequestMessage requestMessage, ResourceType resourceType, CancellationToken cancellationToken) in /home/vsts/work/1/s/Microsoft.Azure.Cosmos/src/HttpClient/CosmosHttpClientCore.cs:line 380
   at Microsoft.Azure.Cosmos.CosmosHttpClientCore.SendHttpHelperAsync(Func`1 createRequestMessageAsync, ResourceType resourceType, HttpTimeoutPolicy timeoutPolicy, ITrace trace, CancellationToken cancellationToken) in /home/vsts/work/1/s/Microsoft.Azure.Cosmos/src/HttpClient/CosmosHttpClientCore.cs:line 252
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Cosmos.ResponseMessage.EnsureSuccessStatusCode() in /home/vsts/work/1/s/Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs:line 161
   at Microsoft.Azure.Cosmos.CosmosResponseFactoryCore.ProcessMessage[T](ResponseMessage responseMessage, Func`2 createResponse) in /home/vsts/work/1/s/Microsoft.Azure.Cosmos/src/Resource/CosmosResponseFactoryCore.cs:line 243
   at Microsoft.Azure.Cosmos.CosmosResponseFactoryCore.CreateItemResponse[T](ResponseMessage responseMessage) in /home/vsts/work/1/s/Microsoft.Azure.Cosmos/src/Resource/CosmosResponseFactoryCore.cs:line 77
   at Microsoft.Azure.Cosmos.ContainerCore.CreateItemAsync[T](T item, ITrace trace, Nullable`1 partitionKey, ItemRequestOptions requestOptions, CancellationToken cancellationToken) in /home/vsts/work/1/s/Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs:line 87
   at Microsoft.Azure.Cosmos.ClientContextCore.RunWithDiagnosticsHelperAsync[TResult](ITrace trace, Func`2 task) in /home/vsts/work/1/s/Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs:line 411
   at Microsoft.Azure.Cosmos.ClientContextCore.OperationHelperWithRootTraceAsync[TResult](String operationName, RequestOptions requestOptions, Func`2 task) in /home/vsts/work/1/s/Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs:line 209
```

Diagnostics

```
{
    "name": "NoOp",
    "id": "00000000-0000-0000-0000-000000000000",
    "component": "Unknown",
    "start time": "12:00:00:000",
    "duration in milliseconds": 0,
    "data": {},
    "children": []
}
```

**Comment by j82w** [comment]
The regression causes all client side request stats to be excluded on exceptions. 

The CosmosClientSideRequestStats is created here, but it is only added the trace if it is a success operation.
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/da197a56b765a1db62517ed6f07077eadc23c2e1/Microsoft.Azure.Cosmos/src/Handler/TransportHandler.cs#L73

It is added to the Trace for success/exceptionless operations here.
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/da197a56b765a1db62517ed6f07077eadc23c2e1/Microsoft.Azure.Cosmos/src/Util/Extensions.cs#L72

If a DocumentClientException is thrown it calls the following function which never adds.
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/da197a56b765a1db62517ed6f07077eadc23c2e1/Microsoft.Azure.Cosmos/src/Util/Extensions.cs#L131

**Comment by j82w** [comment]
Fixed in #2375
